### PR TITLE
Introducing Git backend for Zuul.D

### PR DESCRIPTION
**Pull request by jsanemet**
Adding 'GitBackend', an API that allows Cibyl to retrieve and parse Zuul.D files from a Git repository. Alongside it this PR also adds many tools that the backend uses to do its task. Added 'zuuld.json', a schema for Zuul.D files.

The backend is not integrated, cibyl cannot call it yet.

### CODE REVIEW

File: cibyl/sources/zuuld/errors.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+"""
+
+
+class ZuulDError(Exception):
```
Comment (cescgina) [inline cibyl/sources/zuuld/errors.py:18]: This should inherit from `SourceException` to easily catch errors raised by this source

File: cibyl/sources/zuuld/specs/git.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+@dataclass
+class GitSpec(SCMSpec):
+    """Defines location of Zuul.D files inside a Git repository.
+
+    This class provides some additional checks over its generic counterpart:
+        - For 'remote': It checks that the URL is a valid Git URL. Raises
+        :class:`InvalidURL` if not.
+    """
+
+    def __post_init__(self):
```
Comment (cescgina) [inline cibyl/sources/zuuld/specs/git.py:32]: TIL about `__post_init__` in dataclasses, seems really useful

Comment (jsanemet) [reply cibyl/sources/zuuld/specs/git.py:32]: I didn't know about it either. It's cool for when you want to make some validations after the class gets instanced.

File: cibyl/sources/zuuld/errors.py (hunk 1 above)
Comment (jsanemet) [reply cibyl/sources/zuuld/errors.py:18]: ZuulD is not really a source. A expect that the Zuul source will be catch this kind of error and return them as a SourceException. On any case I will make the change as it won't break anything.

Comment (jsanemet) [reply cibyl/sources/zuuld/errors.py]: Done.

**Review by cescgina — APPROVED** [review]

**Pull request merged**